persistent-postgresql/Database/Persist/Postgresql/Internal/Migration.hs

```diff
@@ -0,0 +1,1107 @@
+                    , crFieldCascade =
+                        FieldCascade
+                            { fcOnUpdate = Just fcOnUpdate
+                            , fcOnDelete = Just fcOnDelete
+                            }
+                    }
+
+        pure $ Map.singleton sourceTableName (Map.singleton sourceColumnName columnRef)
+
+parseCascade :: Text -> Either Text CascadeAction
+parseCascade txt =
+    case txt of
+        "NO ACTION" ->
+            Right NoAction
+        "CASCADE" ->
+            Right Cascade
+        "SET NULL" ->
+            Right SetNull
+        "SET DEFAULT" ->
+            Right SetDefault
+        "RESTRICT" ->
+            Right Restrict
+        _ ->
+            Left $ "Unexpected value in parseCascade: " <> txt
+
+mapLeft :: (a1 -> a2) -> Either a1 b -> Either a2 b
+mapLeft _ (Right x) = Right x
+mapLeft f (Left x) = Left (f x)
+
+-- | Returns a structured representation of all of the
+-- DB changes required to migrate the Entity from its
+-- current state in the database to the state described in
+-- Haskell.
+migrateEntitiesStructured
+    :: (Text -> IO Statement)
+    -> [EntityDef]
+    -> [EntityDef]
+    -> IO (Either [Text] [AlterDB])
+migrateEntitiesStructured getStmt allDefs defsToMigrate = do
+    r <- collectSchemaState getStmt (map getEntityDBName defsToMigrate)
+    pure $ case r of
+        Right schemaState ->
+            migrateEntitiesFromSchemaState schemaState allDefs defsToMigrate
+        Left err ->
+            Left [err]
+
+migrateEntitiesFromSchemaState
+    :: SchemaState
+    -> [EntityDef]
+    -> [EntityDef]
+    -> Either [Text] [AlterDB]
+migrateEntitiesFromSchemaState (SchemaState schemaStateMap) allDefs defsToMigrate =
+    let
+        go :: EntityDef -> Either Text [AlterDB]
+        go entity = do
+            let
+                name = getEntityDBName entity
+            case Map.lookup name schemaStateMap of
+                Just entityState ->
+                    Right $ migrateEntityFromSchemaState entityState allDefs entity
+                Nothing ->
+                    Left $ T.pack $ "No entry for entity in schemaState: " <> show name
+     in
+        case partitionEithers (map go defsToMigrate) of
+            ([], xs) -> Right (concat xs)
+            (errs, _) -> Left errs
+
+migrateEntityFromSchemaState
+    :: EntitySchemaState
+    -> [EntityDef]
+    -> EntityDef
+    -> [AlterDB]
+migrateEntityFromSchemaState schemaState allDefs entity =
+    case schemaState of
+        EntityDoesNotExist ->
+            (addTable newcols entity) : uniques ++ references ++ foreignsAlt
+        EntityExists ExistingEntitySchemaState{essColumns, essConstraints} ->
+            let
+                (acs, ats) =
+                    getAlters
+                        allDefs
+                        entity
+                        (newcols, udspair)
+                        (map dubiouslyRemoveReferences essColumns, Map.toList essConstraints)
```

Comment:
This function is basically the same as before except for this one line - in the previous version, we were relying on `getColumn` to implicitly remove these references (or rather not fetch them in the first place), now we're doing it explicitly here